homeassistant/components/vacuum/template.py

```diff
@@ -0,0 +1,362 @@
+CONF_START_ACTION = 'start'
+CONF_PAUSE_ACTION = 'pause'
+CONF_STOP_ACTION = 'stop'
+CONF_RETURN_TO_BASE_ACTION = 'return_to_base'
+CONF_CLEAN_SPOT_ACTION = 'clean_spot'
+CONF_LOCATE_ACTION = 'locate'
+CONF_SET_FAN_SPEED_ACTION = 'set_fan_speed'
+
+ENTITY_ID_FORMAT = DOMAIN + '.{}'
+_VALID_STATES = [STATE_CLEANING, STATE_DOCKED, STATE_PAUSED, STATE_IDLE,
+                 STATE_RETURNING, STATE_ERROR]
+
+VACUUM_SCHEMA = vol.Schema({
+    vol.Optional(CONF_FRIENDLY_NAME): cv.string,
+    vol.Optional(CONF_VALUE_TEMPLATE): cv.template,
+    vol.Optional(CONF_BATTERY_LEVEL_TEMPLATE): cv.template,
+    vol.Optional(CONF_FAN_SPEED_TEMPLATE): cv.template,
+
+    vol.Required(CONF_START_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_PAUSE_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_STOP_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_RETURN_TO_BASE_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_CLEAN_SPOT_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_LOCATE_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_SET_FAN_SPEED_ACTION): cv.SCRIPT_SCHEMA,
+
+    vol.Optional(
+        CONF_FAN_SPEED_LIST,
+        default=[]
+    ): cv.ensure_list,
+
+    vol.Optional(CONF_ENTITY_ID): cv.entity_ids
+})
+
+PLATFORM_SCHEMA = cv.PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_VACUUMS): vol.Schema({cv.slug: VACUUM_SCHEMA}),
+})
+
+
+async def async_setup_platform(
+        hass, config, async_add_entities, discovery_info=None
+):
+    """Set up the Template Vacuums."""
+    vacuums = []
+
+    for device, device_config in config[CONF_VACUUMS].items():
+        friendly_name = device_config.get(CONF_FRIENDLY_NAME, device)
+
+        state_template = device_config.get(CONF_VALUE_TEMPLATE)
+        battery_level_template = device_config.get(CONF_BATTERY_LEVEL_TEMPLATE)
+        fan_speed_template = device_config.get(CONF_FAN_SPEED_TEMPLATE)
+
+        start_action = device_config[CONF_START_ACTION]
+        pause_action = device_config.get(CONF_PAUSE_ACTION)
+        stop_action = device_config.get(CONF_STOP_ACTION)
+        return_to_base_action = device_config.get(CONF_RETURN_TO_BASE_ACTION)
+        clean_spot_action = device_config.get(CONF_CLEAN_SPOT_ACTION)
+        locate_action = device_config.get(CONF_LOCATE_ACTION)
+        set_fan_speed_action = device_config.get(CONF_SET_FAN_SPEED_ACTION)
+
+        fan_speed_list = device_config[CONF_FAN_SPEED_LIST]
+
+        entity_ids = set()
+        manual_entity_ids = device_config.get(CONF_ENTITY_ID)
+
+        for template in (state_template, battery_level_template,
+                         fan_speed_template):
+            if template is None:
+                continue
+            template.hass = hass
+
+            if entity_ids == MATCH_ALL or manual_entity_ids is not None:
+                continue
+
+            template_entity_ids = template.extract_entities()
+            if template_entity_ids == MATCH_ALL:
+                entity_ids = MATCH_ALL
```

Comment:
If any of the extracted entities from a template `== MATCH_ALL`, the template entity should use manual update mode. Example of how this works: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/sensor/template.py#L76-L88